### PR TITLE
SONARJAVA-1941 fix handling of varargs when applying yields

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

**Comment by m-g-sonar** [comment]
Still need cleaning.

File: java-frontend/src/main/java/org/sonar/java/se/MethodYield.java [hunk 1]
```diff
@@ -36,35 +38,43 @@
...
   }
 
+
   @Override
   public String toString() {
     return "{params: " + Arrays.toString(parametersConstraints) + ", result: " + resultConstraint + " (" + resultIndex + "), exceptional: " + exception + "}";
   }
 
-  public Collection<ProgramState> statesAfterInvocation(List<SymbolicValue> invocationArguments, ProgramState programState, Supplier<SymbolicValue> svSupplier) {
+  public Collection<ProgramState> statesAfterInvocation(List<SymbolicValue> invocationArguments, List<Type> invocationTypes, ProgramState programState,
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/MethodYield.java]: CRITICAL The Cyclomatic Complexity of this method "statesAfterInvocation" is 11 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java [hunk 2]
```diff
@@ -550,10 +550,11 @@ private void executeMethodInvocation(MethodInvocationTree mit) {
     final SymbolicValue resultValue = constraintManager.createMethodSymbolicValue(mit, unstack.values);
     if (methodInvokedBehavior != null && methodInvokedBehavior.isComplete() && methodCanNotBeOverriden(methodSymbol)) {
       List<SymbolicValue> invocationArguments = invocationArguments(unstack.values);
+      List<Type> invocationTypes = mit.arguments().stream().map(t -> t.symbolType()).collect(Collectors.toList());
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java]: MINOR Replace this lambda with a method reference. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1612)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* CRITICAL 1 critical
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* CRITICAL 1 critical

Watch the comments in this conversation to review them.

File: java-frontend/src/main/java/org/sonar/java/se/MethodYield.java [hunk 3]
```diff
@@ -54,12 +59,17 @@ public String toString() {
...
+    Supplier<SymbolicValue> svSupplier) {
     Set<ProgramState> results = new LinkedHashSet<>();
     for (int index = 0; index < invocationArguments.size(); index++) {
-      // FIXME : varargs method should be handled
       SymbolicValue invokedArg = invocationArguments.get(index);
-      Constraint constraint = parametersConstraints[Math.min(index, parametersConstraints.length - 1)];
+      Constraint constraint = null;
+      if (!varArgs || varArgsCalledWithOneArrayParameter(index, invocationTypes)) {
+        constraint = parametersConstraints[index];
+      } else {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/MethodYield.java]: I am not a big fan of an else for the sole purpose of a comment. Rephrase it to make it applicable for the condition ?

File: java-frontend/src/main/java/org/sonar/java/se/MethodYield.java [hunk 4]
```diff
@@ -105,10 +99,28 @@ public String toString() {
...
+      return Optional.ofNullable(parametersConstraints[index]);
+    }
+    return Optional.empty();
+  }
+
+  /**
+   * For varArgs methods, only apply the constraint on single array parameter, in order to not 
+   * wrongly apply it on all the elements of the array.
+   */
+  private boolean appliableOnVarArgs(int index, List<Type> types) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/MethodYield.java]: typo ?

File: java-frontend/src/test/files/se/VarArgsYields.java [hunk 5]
```diff
@@ -1,18 +1,20 @@
 class A {
-  private void varArgMethod(A... objects) {
```
Comment (benzonico) [inline java-frontend/src/test/files/se/VarArgsYields.java]: Seems that you are missing the case where the varargs part is empty. 

varArgMethod(A... objs) called without any arg.

Comment (m-g-sonar) [inline java-frontend/src/test/files/se/VarArgsYields.java]: Nop, it is tested line 14.
